File: main/n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    static vector<int> preorder(const Node* root);
};

vector<int> Solution::preorder(const Node* const root)
{
    vector<int> ret;
    
    stack<const Node*> s;
    s.push(root);
    
    while (!s.empty()) {
        const auto parent = s.top();
        s.pop();
        if (!parent) continue;
        
        ret.push_back(parent->val);
        
        for_each(crbegin(parent->children), crend(parent->children),
                 [&s](const auto child) { s.push(child); });
    }
    
    return ret;
}

File: main/n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> preorder(Node root) {
        final List<Integer> ret = new ArrayList<>();
        if (root == null) return ret;

        final Deque<Node> stack = new ArrayDeque<>();
        stack.push(root);

        while (!stack.isEmpty()) {
            root = stack.pop();
            ret.add(root.val);

            final Deque<Node> children = new ArrayDeque<>(root.children);
            while (!children.isEmpty()) stack.push(children.removeLast());
        }

        return ret;
    }
}
